library/alloc/src/sync.rs

```diff
@@ -2300,6 +2300,19 @@ impl<T: ?Sized + Hash> Hash for Arc<T> {
 
 #[stable(feature = "from_for_ptrs", since = "1.6.0")]
 impl<T> From<T> for Arc<T> {
+    /// Converts a generic type `T` into an `Arc<T>`
+    ///
+    /// The conversion allocates on the heap and moves `t`
+    /// from the stack into it.
```

Comment:
The wording intents to be helpful but is actually to strong:

```
    use std::sync::Arc;
    let x = Box::new(String::new());
    let y: Arc<String> = Arc::from(*x);
```

This moves from the heap to the heap. Suggestion: "and moves `t`into it"

Reply:
If you look at MIR, you can see that dereferencing actually moves it to the stack and then calls `Arc::from`. It might look like it is moving from the heap to the heap, but you have to call it with ownership, and it is on the stack.

Existing documentation already suggests that it moves from the stack. See [this](https://github.com/rust-lang/rust/blob/1969c2e312303faa458cf19dad9783165e3c72c8/library/alloc/src/boxed.rs#L10) as an example. It also says dereferencing moves it back to the stack.

MIR:
```
fn main() -> () {
    let mut _0: ();                      // return place in scope 0 at src/main.rs:2:16: 2:16
    let _1: std::boxed::Box<std::string::String>; // in scope 0 at src/main.rs:3:13: 3:14
    let mut _2: std::string::String;     // in scope 0 at src/main.rs:3:26: 3:39
    let mut _4: std::string::String;     // in scope 0 at src/main.rs:4:40: 4:42
    let mut _5: ();                      // in scope 0 at src/main.rs:5:5: 5:6
    let mut _6: ();                      // in scope 0 at src/main.rs:5:5: 5:6
    scope 1 {
        debug x => _1;                   // in scope 1 at src/main.rs:3:13: 3:14
        let _3: std::sync::Arc<std::string::String>; // in scope 1 at src/main.rs:4:13: 4:14
        scope 2 {
            debug y => _3;               // in scope 2 at src/main.rs:4:13: 4:14
        }
    }

    bb0: {
        _2 = String::new() -> bb1;       // scope 0 at src/main.rs:3:26: 3:39
                                         // mir::Constant
                                         // + span: src/main.rs:3:26: 3:37
                                         // + literal: Const { ty: fn() -> std::string::String {std::string::String::new}, val: Value(Scalar(<ZST>)) }
    }

    bb1: {
        _1 = Box::<String>::new(move _2) -> bb2; // scope 0 at src/main.rs:3:17: 3:40
                                         // mir::Constant
                                         // + span: src/main.rs:3:17: 3:25
                                         // + user_ty: UserType(0)
                                         // + literal: Const { ty: fn(std::string::String) -> std::boxed::Box<std::string::String> {std::boxed::Box::<std::string::String>::new}, val: Value(Scalar(<ZST>)) }
    }

    bb2: {
        _4 = move (*_1);                 // scope 1 at src/main.rs:4:40: 4:42
        _3 = <Arc<String> as From<String>>::from(move _4) -> [return: bb3, unwind: bb7]; // scope 1 at src/main.rs:4:30: 4:43
                                         // mir::Constant
                                         // + span: src/main.rs:4:30: 4:39
                                         // + user_ty: UserType(2)
                                         // + literal: Const { ty: fn(std::string::String) -> std::sync::Arc<std::string::String> {<std::sync::Arc<std::string::String> as std::convert::From<std::string::String>>::from}, val: Value(Scalar(<ZST>)) }
    }

    bb3: {
        _0 = const ();                   // scope 0 at src/main.rs:2:16: 5:6
        drop(_3) -> [return: bb4, unwind: bb7]; // scope 1 at src/main.rs:5:5: 5:6
    }

    bb4: {
        _5 = alloc::alloc::box_free::<String, std::alloc::Global>(move (_1.0: std::ptr::Unique<std::string::String>), move (_1.1: std::alloc::Global)) -> bb5; // scope 0 at src/main.rs:5:5: 5:6
                                         // mir::Constant
                                         // + span: src/main.rs:5:5: 5:6
                                         // + literal: Const { ty: unsafe fn(std::ptr::Unique<std::string::String>, std::alloc::Global) {alloc::alloc::box_free::<std::string::String, std::alloc::Global>}, val: Value(Scalar(<ZST>)) }
    }

    bb5: {
        return;                          // scope 0 at src/main.rs:5:6: 5:6
    }

    bb6 (cleanup): {
        resume;                          // scope 0 at src/main.rs:2:5: 5:6
    }

    bb7 (cleanup): {
        _6 = alloc::alloc::box_free::<String, std::alloc::Global>(move (_1.0: std::ptr::Unique<std::string::String>), move (_1.1: std::alloc::Global)) -> bb6; // scope 0 at src/main.rs:5:5: 5:6
                                         // mir::Constant
                                         // + span: src/main.rs:5:5: 5:6
                                         // + literal: Const { ty: unsafe fn(std::ptr::Unique<std::string::String>, std::alloc::Global) {alloc::alloc::box_free::<std::string::String, std::alloc::Global>}, val: Value(Scalar(<ZST>)) }
    }
}
```

Reply:
I don't think we should be documenting that it comes 'from the stack' here. It might come out of a register or a hardcoded constant in the instructions, etc. And things might change in a future smarter version of the compiler.

Since this simply calls `Arc::new`, we should probably just document that. Maybe "moves the value into a newly allocated Arc" if you want to specifically mention the allocation.

Reply:
I took my documentation from `Box`.
Look at this:
https://github.com/rust-lang/rust/blob/5ea19239d9d6f49fdd76513a36386d7e83708e3f/library/alloc/src/boxed.rs#L1215